File: POProjectJAVA/asteroids/Menu.java
package asteroids;

import java.awt.Canvas;
import java.awt.Dimension;
import java.awt.Graphics;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;
import java.awt.event.KeyEvent;
import java.awt.event.KeyListener;
import java.awt.image.BufferStrategy;
import java.awt.event.WindowEvent;
import javax.swing.JFrame;

class Menu extends Canvas implements Runnable, MouseListener, KeyListener {
  private static final long serialVersionUID = 1L;
  private String name;
  private boolean running;
  private int windowSize;
  private int segments;
  private int money;
  private int framerate;
  private Theme theme;
  private String tab;
  private Game game;
  private Store store;
  private JFrame window;
  private int gridTile;
  private Draw draw;

  public Menu() {
    this.name = "Asteroids";
    this.running = true;
    this.tab = "mainMenu";
    this.windowSize = 900;
    this.segments = 30;
    this.gridTile = this.windowSize / this.segments;
    this.money = 70;
    this.framerate = 60;
    this.window = new JFrame(this.name);
    this.store = new Store();
    this.game = new Game(this.store.getCurrentItem(), 2);
    this.theme = new Theme();
    this.draw = new Draw(this.theme, this.windowSize, this.gridTile);
    this.window.addMouseListener(this);
    setMinimumSize(new Dimension(this.windowSize, this.windowSize));
    setMaximumSize(new Dimension(this.windowSize, this.windowSize));
    setPreferredSize(new Dimension(this.windowSize, this.windowSize));
    addMouseListener(this);
    addKeyListener(this);

    this.window.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);

    this.window.add(this);
    this.window.pack();

    this.window.setResizable(false);
    this.window.setLocationRelativeTo(null);
    this.window.setVisible(true);
  }

  public synchronized void start() {
    this.running = true;
    new Thread(this).start();
  }

  public synchronized void stop() {
    this.running = false;
    this.window.dispatchEvent(new WindowEvent(this.window, WindowEvent.WINDOW_CLOSING));
  }

  void startNewGame(int difficulty) {
    this.game = new Game(this.store.getCurrentItem(), difficulty);
    this.tab = "game";
  }

  public void run() {
    long lastTime = System.nanoTime();
    double delta = 0;

    while (running) {
      long now = System.nanoTime();
      delta += (now - lastTime) / (1000000000D / (double) this.framerate);
      lastTime = now;
      boolean gameFrame = false;
      while (delta >= 1) {
        delta -= 1;
        gameFrame = true;
      }
      if (gameFrame) {
        this.update();
        this.draw();
      }
    }
  }

  public void update() {
    if (this.tab == "game") {
      this.game.update();
      if (this.game.isOver()) {
        this.money += game.getScore();
        this.tab = "endGame";
      }
    }
  }

  public void draw() {
    BufferStrategy bs = getBufferStrategy();
    if (bs == null) {
      createBufferStrategy(3);
      return;
    }
    Graphics g = bs.getDrawGraphics();
    if (this.tab == "mainMenu")
      this.draw.mainMenu(g);
    else if (this.tab == "difficultyMenu")
      this.draw.difficultyMenu(g);
    else if (this.tab == "game")
      this.draw.game(g, this.game);
    else if (this.tab == "endGame")
      this.draw.endGame(g, this.game);
    else if (this.tab == "store")
      this.draw.store(g, this.store, this.money);
    g.dispose();
    bs.show();
  }

  public void handleMouseInput(int x, int y) {
    if (this.tab == "mainMenu") {
      if (x >= 6 && x <= 24 && y >= 12 && y <= 16)
        this.tab = "difficultyMenu";
      if (x >= 6 && x <= 24 && y >= 18 && y <= 22)
        this.tab = "store";
      if (x >= 6 && x <= 24 && y >= 24 && y <= 28)
        this.stop();
    }

    else if (this.tab == "difficultyMenu") {
      if (x >= 6 && x <= 24 && y >= 6 && y <= 10)
        this.startNewGame(1);
      if (x >= 6 && x <= 24 && y >= 12 && y <= 16)
        this.startNewGame(2);
      if (x >= 6 && x <= 24 && y >= 18 && y <= 22)
        this.startNewGame(3);
      if (x >= 6 && x <= 24 && y >= 24 && y <= 28)
        this.startNewGame(4);
    }

    else if (this.tab == "endGame") {
      if (x >= 8 && x <= 22 && y >= 16 && y <= 19)
        this.startNewGame(this.game.getDifficulty());
      if (x >= 8 && x <= 22 && y >= 21 && y <= 24)
        this.tab = "mainMenu";
    }

    else if (this.tab == "store") {
      if (x >= 6 && x <= 24 && y >= 24 && y <= 28)
        this.tab = "mainMenu";
      else if (x >= 3 && x <= 9 && y >= 13 && y <= 16)
        this.money -= this.store.selectItem(0, this.money);
      else if (x >= 12 && x <= 18 && y >= 13 && y <= 16)
        this.money -= this.store.selectItem(1, this.money);
      else if (x >= 21 && x <= 27 && y >= 13 && y <= 16)
        this.money -= this.store.selectItem(2, this.money);
    }

    else if (this.tab == "game" && this.game.getPaused()) {
      if (x >= 8 && x <= 22 && y >= 11 && y <= 14)
        this.game.setPaused(false);
      if (x >= 8 && x <= 22 && y >= 16 && y <= 19)
        this.tab = "mainMenu";
    }
  }

  @Override
  public void mouseClicked(MouseEvent e) {
    int x = e.getX();
    int y = e.getY();
    this.handleMouseInput(x / this.gridTile, y / this.gridTile);
  }

  @Override
  public void mouseExited(MouseEvent e) {
  }

  @Override
  public void mouseEntered(MouseEvent e) {
  }

  @Override
  public void mouseReleased(MouseEvent e) {
  }

  @Override
  public void mousePressed(MouseEvent e) {
  }

  @Override
  public void keyTyped(KeyEvent e) {

  }

  @Override
  public void keyPressed(KeyEvent e) {
    this.game.handleInput(e.getKeyCode(), true);
  }

  @Override
  public void keyReleased(KeyEvent e) {
    this.game.handleInput(e.getKeyCode(), false);
  }

  public static void main(String args[]) {
    Menu menu = new Menu();
    menu.run();
  }
}

File: POProjectCPP/Menu.cpp
#include "Menu.h"

/*
 _____ _____ _   _  _____ ___________ _   _ _____ _____ ___________  _____ 
/  __ \  _  | \ | |/  ___|_   _| ___ \ | | /  __ \_   _|  _  | ___ \/  ___|
| /  \/ | | |  \| |\ `--.  | | | |_/ / | | | /  \/ | | | | | | |_/ /\ `--. 
| |   | | | | . ` | `--. \ | | |    /| | | | |     | | | | | |    /  `--. \
| \__/\ \_/ / |\  |/\__/ / | | | |\ \| |_| | \__/\ | | \ \_/ / |\ \ /\__/ /
 \____/\___/\_| \_/\____/  \_/ \_| \_|\___/ \____/ \_/  \___/\_| \_|\____/ 
                                                                           
*/

Menu::Menu(int windowSize, sf::RenderWindow *window) : window(nullptr)
{
  this->tab = "mainMenu";
  this->windowSize = windowSize;
  this->segments = 30;
  this->money = 70;
  this->window = window;
  this->store = Store();
  this->drawing = Draw(window, &(store.getSelectedProduct()), windowSize, this->segments);
  this->game = Game(this->segments);
  this->setFramerate(10);
  this->selectTheme(0);
}

/*
 _____  _____ _____ _____ ___________  _____            _____ _____ _____ _____ ___________  _____ 
|  __ \|  ___|_   _|_   _|  ___| ___ \/  ___|   ___    /  ___|  ___|_   _|_   _|  ___| ___ \/  ___|
| |  \/| |__   | |   | | | |__ | |_/ /\ `--.   ( _ )   \ `--.| |__   | |   | | | |__ | |_/ /\ `--. 
| | __ |  __|  | |   | | |  __||    /  `--. \  / _ \/\  `--. \  __|  | |   | | |  __||    /  `--. \
| |_\ \| |___  | |   | | | |___| |\ \ /\__/ / | (_>  < /\__/ / |___  | |   | | | |___| |\ \ /\__/ /
 \____/\____/  \_/   \_/ \____/\_| \_|\____/   \___/\/ \____/\____/  \_/   \_/ \____/\_| \_|\____/ 
    
*/

sf::RenderWindow *Menu::getWindow()
{
  return this->window;
}

int Menu::getFramerate()
{
  return this->framerate;
}

void Menu::setFramerate(int framerate)
{
  this->framerate = framerate;
  this->window->setFramerateLimit(framerate);
}

void Menu::selectTheme(int index)
{
  this->store.selectItem(index, this->money);
}

/*
 _     _____ _____ _____ _____ 
| |   |  _  |  __ \_   _/  __ \
| |   | | | | |  \/ | | | /  \/
| |   | | | | | __  | | | |    
| |___\ \_/ / |_\ \_| |_| \__/\
\_____/\___/ \____/\___/ \____/
        
*/

void Menu::handleKeybordInput(sf::Event event)
{
  if (this->tab == "game")
    this->game.handleKeybordInput(event);
}

void Menu::handleMouseInput(sf::Event event)
{
  float tileSize = this->windowSize / this->segments;
  int x = event.mouseButton.x / tileSize;
  int y = event.mouseButton.y / tileSize;
  if (this->tab == "mainMenu")
  {
    if (x >= 6 && x <= 24 && y >= 12 && y <= 16)
      this->tab = "difficultyMenu";
    if (x >= 6 && x <= 24 && y >= 18 && y <= 22)
      this->tab = "store";
    if (x >= 6 && x <= 24 && y >= 24 && y <= 28)
      this->window->close();
  }
  else if (this->tab == "difficultyMenu")
  {
    if (x >= 6 && x <= 24 && y >= 6 && y <= 10)
    {
      this->setFramerate(10);
      this->game.restart();
      this->tab = "game";
    }
    if (x >= 6 && x <= 24 && y >= 12 && y <= 16)
    {
      this->setFramerate(15);
      this->game.restart();
      this->tab = "game";
    }
    if (x >= 6 && x <= 24 && y >= 18 && y <= 22)
    {
      this->setFramerate(20);
      this->game.restart();
      this->tab = "game";
    }
    if (x >= 6 && x <= 24 && y >= 24 && y <= 28)
    {
      this->setFramerate(25);
      this->game.restart();
      this->tab = "game";
    }
  }
  else if (this->tab == "store")
  {
    for (int i = 0; i < this->store.getItems().size(); i++)
    {
      if (x >= (3 + 9 * i) && x <= (9 + 9 * i) && y >= 13 && y <= 16)
      {
        this->selectTheme(i);
      }
    }
    if (x >= 6 && x <= 24 && y >= 24 && y <= 28)
      this->tab = "mainMenu";
  }
  else if (this->tab == "endGame")
  {
    if (x >= 8 && x <= 22 && y >= 16 && y <= 19)
    {
      this->game.restart();
      this->tab = "game";
    }
    if (x >= 8 && x <= 22 && y >= 21 && y <= 24)
      this->tab = "mainMenu";
  }
  else if (this->tab == "game" && this->game.getPaused())
  {
    if (x >= 8 && x <= 22 && y >= 11 && y <= 14)
    {
      this->game.setPaused(false);
    }
    if (x >= 8 && x <= 22 && y >= 16 && y <= 19)
      this->tab = "mainMenu";
  }
}

void Menu::update()
{
  if (this->tab == "game")
  {
    this->game.update();
    if (this->game.isOver())
    {
      this->tab = "endGame";
      this->money += this->game.getScore();
    }
  }
}

void Menu::draw()
{
  this->drawing.setTheme(&(store.getSelectedProduct()));
  this->drawing.background();
  if (this->tab == "mainMenu")
    this->drawing.mainMenu();
  if (this->tab == "difficultyMenu")
    this->drawing.difficultyMenu();
  if (this->tab == "game")
    this->drawing.game(this->game);
  if (this->tab == "endGame")
    this->drawing.endGame(this->game);
  if (this->tab == "store")
    this->drawing.store(this->store, this->money);
}